File: models/Hospitalisations/FNETS/backtest.R
# Backtest FNETS model on hospital data.

#!/usr/bin/env Rscript
# USAGE: Rscript backtest.R  backtesting_config.yaml <DESIGN_MATRIX>.csv 

library("fnets") 

# Check if the correct number of arguments is provided
if (length(commandArgs(trailingOnly = TRUE)) != 2) {
    stop("USAGE: Rscript FNETS-backtest.R backtesting_config.yaml <DESIGN_MATRIX>.csv")
}

# Get the config file path from the second argument
config_file <- commandArgs(trailingOnly = TRUE)[1]

# Read the YAML file
yaml_data <- yaml::read_yaml(config_file)

# Access individual variables and assign them to R variables
SEED <- yaml_data$SEED
n_backtest_days_total <- yaml_data$n_backtest_days
first_prediction_day <- yaml_data$first_prediction_day
Q <- yaml_data$Q 
target_feature <- yaml_data$target_feature 
lookback_window <- yaml_data$lookback_window

###### ENVIRONMENT VARIABLES ###### 
# PBS_ARRAY_INDEX <- as.numeric(Sys.getenv("PBS_ARRAY_INDEX"))
# NUM_ARRAY_INDICES <- as.numeric(Sys.getenv("NUM_ARRAY_INDICES")) 
PBS_ARRAY_INDEX = 1
NUM_ARRAY_INDICES = 1

# Re-define n_backtest_days to be total number of backtesting days divided by the number of array indices 
n_backtest_days <- as.integer(n_backtest_days_total/NUM_ARRAY_INDICES) 
print(n_backtest_days_total)

# Get a list of days to do backtesting on
days_to_backtest <- 1:n_backtest_days + (n_backtest_days * (PBS_ARRAY_INDEX - 1)) - 1 

# Get the design matrix file path from the second argument
design_matrix_file <- commandArgs(trailingOnly = TRUE)[2]

# Read the design matrix
Xs <- read.csv2(design_matrix_file,sep=",",dec = ".",header=TRUE)

# Get dimensions
T <- nrow(Xs)
N_times_Q <- ncol(Xs)

N <- N_times_Q / Q

s <- list()

factors <- list()

for (index in 1:n_backtest_days){
    t <- days_to_backtest[[index]] 
    todays_date <- first_prediction_day + t
    print(todays_date)
    furthest_lookback_day <- todays_date - lookback_window
    X_train <- Xs[1:(todays_date),] 
    X_train <- data.matrix(X_train)
    print(dim(X_train))
    # X_train_diff <- diff(X_train) 
    fit_fnets <- fnets(x=X_train,
                    center = FALSE,
                    fm.restricted = TRUE,
                    q = 1,
                    var.order = 1,
                    var.method = "lasso",
                    do.threshold = FALSE,
                    do.lrpc = FALSE,
                    ) 
    if (fit_fnets$q == 0){ 
      fit_fnets <- fnets(x=X_train,
                    center = FALSE,
                    fm.restricted = TRUE,
                    q = 1,
                    var.order = 1,
                    var.method = "lasso",
                    do.threshold = FALSE,
                    do.lrpc = FALSE,
                    ) 
    }
    print(fit_fnets$q)
    print(object.size(fit_fnets)) 
    factors[[index]] <- fit_fnets$q 
    pr <- predict(fit_fnets, n.ahead = 1,fc.restricted = TRUE)
    predictions_matrix <- pr$forecast  
    # Check if all elements in the matrix are 0
    if(all(predictions_matrix == 0)) {
      print("The predictions matrix contains all 0s.") 
      print(fit_fnets) 
    }  
    # s[[index]] <- predictions_matrix[1:N] + X_train[(lookback_window + index - 1),]
    s[[index]] <- predictions_matrix[1:N] 
    rm(fit_fnets)
}

###### SAVE TO FILE ######
predictions_path <- sprintf("predictions-%d.csv", PBS_ARRAY_INDEX)
s_matrix <- do.call(rbind, s)
write.table(s_matrix, file = predictions_path, sep = ",", row.names = FALSE, col.names = FALSE, quote = FALSE)

factors_path <- sprintf("factors-%d.csv", PBS_ARRAY_INDEX)
f_matrix <- do.call(rbind, factors) 
write.table(f_matrix, file = factors_path, sep = ",", row.names = FALSE, col.names = FALSE, quote = FALSE)

# Write configuration variables to a text file
output_file <- "backtesting_hyp.txt"
f <- file(output_file, "w")

writeLines("{", f)
writeLines(paste0("'SEED':'", SEED, "'"), f)
writeLines(paste0("'n_backtest_days_total':'", n_backtest_days_total, "'"), f)
writeLines(paste0("'first_prediction_day':'", first_prediction_day, "'"), f)
writeLines(paste0("'Q':'", Q, "'"), f)
writeLines(paste0("'target_feature':'", target_feature, "'"), f)
writeLines("}", f)

close(f)

File: models/FinancialReturns/FARM/backtest.py
""" 
Do FARM returns prediction using my implementation of the FARM model.
"""

#!/usr/bin/env python3 
# USAGE: ./backtest.py <DESIGN_MATRIX>.csv backtesting_config.yaml 

import sys
import yaml
import numpy as np
from numpy.random import default_rng
import fcntl
import os 
from sklearn.preprocessing import MinMaxScaler
from scipy.sparse.linalg import eigs
import pandas as pd
from sklearn.linear_model import LinearRegression
from sklearn.linear_model import LassoCV
from sklearn.linear_model import Lasso

with open(sys.argv[2], "r") as f:
    config = yaml.load(f, Loader=yaml.FullLoader)

###### CONFIG PARAMETERS ###### 
SEED = config['SEED']
n_backtest_days_tot = config['n_backtest_days']
first_prediction_day = config['first_prediction_day']
num_lags = config['num_lags']
target_feature = config['target_feature']
Q = config['Q']
lookback_window = config['lookback_window']

###### ENVIRONMENT VARIABLES ###### 
# PBS_ARRAY_INDEX = int(os.environ['PBS_ARRAY_INDEX'])
PBS_ARRAY_INDEX = 1
# NUM_ARRAY_INDICES = int(os.environ['NUM_ARRAY_INDICES'])
NUM_ARRAY_INDICES = 1

# Re-define n_backtest_days to be total number of backtesting days divided by the number of array indices 
n_backtest_days = int(n_backtest_days_tot/NUM_ARRAY_INDICES)

# Get a list of days to do backtesting on
days_to_backtest = [int(i + (n_backtest_days)*(PBS_ARRAY_INDEX-1)) for i in range(n_backtest_days)]

random_state = default_rng(seed=SEED)

###### READ IN DATA ######
Xs = np.genfromtxt(sys.argv[1], delimiter=',')
T = Xs.shape[0]
N_times_Q = Xs.shape[1]
N = N_times_Q/Q
if N != int(N):
    print("ERROR:Input is not a whole number")
N = int(N)

Xs = np.reshape(Xs,(T,N,Q),order='F') 

###### BACKTESTING ###### 
s_array = np.zeros((n_backtest_days,N)) #predictions 

for index in range(n_backtest_days):
    t = days_to_backtest[index]
    print(t)
    todays_date = first_prediction_day+t
    furthest_lookback_day = todays_date - lookback_window
    X_train = Xs[furthest_lookback_day:todays_date,:,target_feature] #Shape = (lookback_window,N)

    # Compute factor and idiosyncratic design matrices 
    evals, evecs = eigs(X_train.T@X_train, k = 1) 
    evals = evals.real 
    evecs = evecs.real
    loadings = np.multiply(np.sqrt(evals),evecs)
    factors = X_train@loadings@np.linalg.inv(np.diag(evals)) 
    U = X_train - factors@loadings.T 
    # Factor regression OLS
    factor_design = np.zeros(((lookback_window-num_lags),num_lags))
    for tau in range(lookback_window-num_lags): 
        covariate_tau = factors[tau:tau+num_lags].flatten()
        factor_design[tau] = covariate_tau
    factor_targets = factors[num_lags:T] 
    ols_reg_object = LinearRegression(fit_intercept=False)
    ols_fit = ols_reg_object.fit(factor_design,factor_targets) 
    factor_coef = ols_fit.coef_  
    # U regression LASSO
    idiosyncratic_design = np.zeros((lookback_window-num_lags,N*num_lags))
    for l in range(lookback_window-num_lags): 
        covariate_l = U[l:l+num_lags].flatten() 
        idiosyncratic_design[l] = covariate_l 
    idiosyncratic_targets = U[num_lags:lookback_window,:] 
    # lasso_reg_object = LassoCV(alphas=[0.0001,0.0005,0.001,0.0015,0.002,0.005,0.01,0.015,0.02,0.05],fit_intercept=False)
    lasso_reg_object = Lasso(alpha = 0.00005,fit_intercept=False)
    lasso_fit = lasso_reg_object.fit(idiosyncratic_design,idiosyncratic_targets) 
    idio_coef = lasso_fit.coef_  
    print(idio_coef[idio_coef!=0].size)

    #predict next day returns 
    estimated_factor = factor_coef@factors[-num_lags:] 
    estimated_idio = idio_coef.T@U[lookback_window-num_lags:].flatten()
    s = estimated_factor*loadings + estimated_idio[:,None]
    s_array[index] = s[:,0]

    print ("\033[A                             \033[A") 
    sys.stdout.write("\033[F")  # \033[F is the ANSI escape sequence to move up one line
    sys.stdout.write("\033[K")  # \033[K is the ANSI escape sequence to clear the line from cursor to the end


###### OUTPUT s_array, labels_hat AND phi_hat TO FILE ###### 
predictions_path = f"predictions-{PBS_ARRAY_INDEX}.csv"


np.savetxt(predictions_path, s_array, delimiter=',', fmt='%.6f')

###### OUTPUT BACKTESTING HYPERPARAMETERS TO FILE ######

f = open("backtesting_hyp.txt", "w")
f.write("{\n")
for k in config.keys():
    f.write("'{}':'{}'\n".format(k, config[k]))
f.write("}")
f.close()